### PR TITLE
line adjacency 

### PR DESCRIPTION
**Pull request by petersteneteg**
Extended the mesh shader cache to support more kinds of state. Made the line renderer handle adjacency property by inspecting the input data.

### CODE REVIEW

File: modules/basegl/src/datastructures/meshshadercache.cpp [hunk 1]
```diff
@@ -37,53 +37,48 @@ namespace inviwo {
...
+            return -1;
+        }
+    }}
+    , updateShader{[glslType, name = name.empty() ? toString(bufferType) : name](
+                       int location, Shader& shader) -> void {
+        if (location >= 0) {
+            const auto& buffername = name;
+            shader[ShaderType::Vertex]->addInDeclaration("in_" + buffername, location, glslType);
+            for (auto& obj : shader.getShaderObjects()) {
+                obj.addShaderDefine("HAS_" + toUpper(buffername));
```
Comment (ResearchDaniel) [inline modules/basegl/src/datastructures/meshshadercache.cpp:57]: Defines never removed?

File: modules/basegl/src/processors/linerenderer.cpp [hunk 2]
```diff
@@ -80,8 +68,16 @@ LineRenderer::LineRenderer()
...
-
+           {BufferType::PickingAttrib, MeshShaderCache::Optional, "uint"},
+           {[](const Mesh&, Mesh::MeshInfo mi) -> int {
+                return mi.ct == ConnectivityType::Adjacency ||
+                               mi.ct == ConnectivityType::StripAdjacency
+                           ? 1
+                           : 0;
+            },
+            [](int mode, Shader& shader) {
+                shader[ShaderType::Geometry]->addShaderDefine("ENABLE_ADJACENCY", toString(mode));
```
Comment (ResearchDaniel) [inline modules/basegl/src/processors/linerenderer.cpp:79]: Never removed?

**Review by ResearchDaniel — APPROVED** [review]
Looks good. Might be an issue with defines added but never removed

File: modules/basegl/src/datastructures/meshshadercache.cpp (hunk 1 above)
Comment (petersteneteg) [reply modules/basegl/src/datastructures/meshshadercache.cpp:57]: It only does run that when the shader is created, if a different value is needed a different shader is created. Hence it should never remove anything

File: modules/basegl/src/processors/linerenderer.cpp (hunk 2 above)
Comment (petersteneteg) [reply modules/basegl/src/processors/linerenderer.cpp:79]: see above

File: modules/basegl/src/processors/linerenderer.cpp [hunk 3]
```diff
@@ -146,53 +130,36 @@ void LineRenderer::process() {
...
-        utilgl::setShaderUniforms(shader, *elem, "geometry");
-        if (autoDrawMode) {
-            drawer.draw();
+    for (const auto& mesh : inport_) {
+        if (mesh->getNumberOfBuffers() == 0) continue;
+        MeshDrawerGL::DrawObject drawer(mesh->getRepresentation<MeshGL>(),
+                                        mesh->getDefaultMeshInfo());
+        if (mesh->getNumberOfIndicies() > 0) {
+            for (std::size_t i = 0; i < mesh->getNumberOfIndicies(); ++i) {
+                if (mesh->getIndexMeshInfo(i).dt != DrawType::Lines) continue;
```
Comment (martinfalk) [inline modules/basegl/src/processors/linerenderer.cpp:139]: so one has to explicitly convert a tri mesh into lines before using this processor?

**Review by martinfalk — APPROVED** [review]

File: modules/basegl/src/processors/linerenderer.cpp (hunk 3 above)
Comment (petersteneteg) [reply modules/basegl/src/processors/linerenderer.cpp:139]: Yes. I think we should optimise for the standard use case.

**Pull request merged**